app/models/user_auth_token.rb

```diff
@@ -30,6 +30,40 @@ def self.log(info)
     end
   end
 
+  # Returns the login location as it will be used by the the system to detect
+  # suspicious login.
+  #
+  # This should not be very specific because small variations in location
+  # (i.e. changes of network, small trips, etc) will be detected as suspicious
+  # logins.
+  #
+  # On the other hand, if this is too broad it will not report any suspicious
+  # logins at all.
+  #
+  # For example, let's choose the country as the only component in login
+  # locations. In general, this should be a pretty good choce with the
+  # exception that for users from huge countries it might not be specific
+  # enoguh. For US users where the real user and the malicious one could
+  # happen to live both in USA, this will not detect any suspicious activity.
+  def self.login_location(ip)
+    DiscourseIpInfo.get(ip)[:country]
+  end
+
+  def self.is_suspicious(user_id, user_ip)
+    ips = UserAuthTokenLog.where(user_id: user_id).pluck('DISTINCT client_ip')
+    return false if ips.empty? # first login is never suspicious
+
+    user_location = login_location(user_ip)
+
+    ips.each do |ip|
```

Comment:
This can be a one-liner ;)

```suggestion
    ips.any? { |ip| user_location == login_location(ip) }
```

Reply:
Is actually `.none?`

Thanks! To my shame, I did not know that `.any?` can receive a block.

Reply:
No shame in that ;)